src/index.js

```diff
@@ -1,101 +1,140 @@
   return String(addr.decapsulate('p2p-stardust'))
 }
 
+function noop () { }
+
 /**
-  * Stardust Transport
-  * @class
-  * @param {Object} options - Options for the listener
-  * @param {PeerId} options.id - Id for the crypto challenge
-  * @param {Transport[]} options.transports - Transport(s) for microswitch
-  * @param {Muxer[]} options.muxers - Muxer(s) for microswitch
-  * @param {boolean} options.softFail - Whether to softly fail on listen errors
-  */
+* Stardust Transport
+* @class
+*/
 class Stardust {
-  constructor ({ transports, muxers, id, softFail }) {
-    this.switch = new MicroSwitch({ transports, addresses: [], muxers })
+  /**
+   * @constructor
+   * @param {Object} options - Options for the listener
+   * @param {Upgrader} options.upgrader
+   * @param {PeerId} options.id - Id for the crypto challenge
+   * @param {Libp2p} options.libp2p - Libp2p instance.
+   * @param {Transport[]} options.transports - Transport(s) for microswitch
+   * @param {Muxer[]} options.muxers - Muxer(s) for microswitch
+   * @param {boolean} options.softFail - Whether to softly fail on listen errors
+   */
+  constructor ({ upgrader, libp2p = {}, id, softFail }) {
+    assert(upgrader, 'An upgrader must be provided. See https://github.com/libp2p/interface-transport#upgrader.')
+    this._upgrader = upgrader
+
+    this.libp2p = libp2p
     this.id = id
     this.softFail = softFail
 
-    this.discovery = new EE()
+    this.connections = {}
+
+    // Discovery
+    this.discovery = new EventEmitter()
     this.discovery.tag = 'stardust'
-    this.discovery.enabled = false
-    this.discovery.start = (callback) => {
-      this.discovery.enabled = true
-      setImmediate(callback)
+    this.discovery._isStarted = false
+    this.discovery.start = () => {
+      this.discovery._isStarted = true
     }
-    this.discovery.stop = (callback) => {
-      this.discovery.enabled = false
-      setImmediate(callback)
+    this.discovery.stop = () => {
+      this.discovery._isStarted = false
     }
+  }
 
-    this.connections = {}
+  /**
+   * @async
+   * @param {Multiaddr} ma
+   * @param {object} options
+   * @param {AbortSignal} options.signal Used to abort dial requests
+   * @returns {Connection} An upgraded Connection
+   */
+  async dial (ma, options = {}) {
+    const stream = await this._connect(ma, options)
+    const maConn = toConnection({
+      stream,
+      remoteAddr: ma
+    }, options)
+    log('new outbound connection %s', maConn.remoteAddr)
+    const conn = await this._upgrader.upgradeOutbound(maConn)
+    log('outbound connection %s upgraded', maConn.remoteAddr)
+    return conn
   }
 
-  dial (ma, options, callback) {
-    if (isFunction(options)) {
-      callback = options
-      options = {}
+  /**
+   * @private
+   * @param {Multiaddr} ma
+   * @param {object} options
+   * @param {AbortSignal} options.signal Used to abort dial requests
+   * @returns {Promise<SimplePeer>} Resolves a SimplePeer Webrtc channel
```

Comment:
webrtc. sounds wrong...

Reply:
yes, that's what happen when you copy js docs 😅